### PR TITLE
iio:iio_ad7124: initial commit

### PR DESCRIPTION
**Pull request by adrimbarean**
Add AD7124 IIO driver. Closely resembles the Linux driver to maintain
compatibility with tools that use it:
https://github.com/analogdevicesinc/linux/blob/master/drivers/iio/adc/ad7124.c

Do not merge before https://github.com/analogdevicesinc/no-OS/pull/904

Signed-off-by: Andrei Drimbarean <Andrei.Drimbarean@analog.com>

### CODE REVIEW

File: iio/iio_ad7124/iio_ad7124.c [hunk 1]
```diff
@@ -0,0 +1,1061 @@
...
+	return bytes_count;
+}
+
+/**
+ * @brief Initialize the AD7124 IIO driver.
+ * @param [out] desc - Pointer to the driver handler.
+ * @param [in] init_param - Pointer to the initialization structure.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t iio_ad7124_init(struct iio_ad7124_desc **desc,
```
Comment (mchindri) [inline iio/iio_ad7124/iio_ad7124.c:1011]: This function is unnecesary.
Create a structure like: https://github.com/analogdevicesinc/no-OS/blob/5a93b5a6e9fb132fd92a1bef0050eb27d5e32aa8/iio/iio_demo/iio_demo_dev.h#L145

File: iio/iio_ad7124/iio_ad7124.c [hunk 2]
```diff
@@ -0,0 +1,1061 @@
...
+}
+
+/**
+ * @brief Transfer data from device into RAM.
+ * @param iio_inst - Physical instance of a iio_axi_adc device.
+ * @param bytes_count - Number of bytes to transfer.
+ * @param ch_mask - Opened channels mask.
+ * @return bytes_count or negative value in case of error.
+ */
+static ssize_t iio_ad7124_transfer_dev_to_mem(void *iio_inst,
```
Comment (mchindri) [inline iio/iio_ad7124/iio_ad7124.c:912]: Did you tried: https://github.com/analogdevicesinc/no-OS/blob/5a93b5a6e9fb132fd92a1bef0050eb27d5e32aa8/iio/iio_demo/iio_demo_dev.h#L151

Comment (mchindri) [reply iio/iio_ad7124/iio_ad7124.c:912]: and read_dev functions

File: iio/iio_ad7124/iio_ad7124.c [hunk 3]
```diff
@@ -0,0 +1,1061 @@
...
+	NULL
+};
+
+#define AD7124_IIO_CHANN_DEF(nm, ch1, ch2, chi) \
+	struct iio_channel ad7124_ch_##chi = { \
+		.name = nm, \
+		.ch_type = IIO_VOLTAGE, \
+		.channel = ch1, \
+		.channel2 = ch2, \
+		.scan_type = NULL, \
```
Comment (mchindri) [inline iio/iio_ad7124/iio_ad7124.c:143]: You should set a scan_type

**Review by mchindri — CHANGES_REQUESTED** [review]

File: iio/iio_ad7124/iio_ad7124.c [hunk 4]
```diff
@@ -0,0 +1,1061 @@
...
+		return FAILURE;
+
+	ldesc->iio_ad7124_device = calloc(1,
+					  sizeof(*ldesc->iio_ad7124_device));
+	if (!ldesc->iio_ad7124_device) {
+		free(ldesc);
+		return -ENOMEM;
+	}
+
+	ldesc->iio_ad7124_device->num_ch = 8;
```
Comment (cristiantpop) [inline iio/iio_ad7124/iio_ad7124.c:1027]: ARRAY_SIZE(ad7124_channels)

File: iio/iio_ad7124/iio_ad7124.c [hunk 5]
```diff
@@ -0,0 +1,1061 @@
...
+ * @brief Set ODR of the device.
+ * @param [in] dev - Pointer to the application handler.
+ * @param [in] odr - New ODR of the device.
+ * @param [in] ch_no - Channel number.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+static float _ad7124_iio_set_odr(struct ad7124_dev *dev, float odr,
+				 int16_t ch_no)
+{
+
```
Comment (cristiantpop) [inline iio/iio_ad7124/iio_ad7124.c:677]: extra line

File: iio/iio_ad7124/iio_ad7124.c [hunk 6]
```diff
@@ -0,0 +1,1061 @@
...
+ * @param channel - IIO channel information.
+ * @return Number of bytes printed in the output buffer, or negative error code.
+ */
+static ssize_t ad1724_iio_read_error(void *device, char *buf, size_t len,
+				     const struct iio_ch_info *channel)
+{
+	struct ad7124_dev	*desc = (struct ad7124_dev *)device;
+	uint32_t		value;
+
+	ad7124_read_register(desc, &desc->regs[AD7124_ERR_REG]);
```
Comment (cristiantpop) [inline iio/iio_ad7124/iio_ad7124.c:484]: Check return value for function calls

File: iio/iio_ad7124/iio_ad7124.c [hunk 7]
```diff
@@ -0,0 +1,1061 @@
...
+ */
+static float _ad7124_iio_get_odr(struct ad7124_dev *dev, int16_t ch_no)
+{
+	float f_clk;
+	struct ad7124_st_reg *reg_map;
+	uint16_t fs_value, flt_coff;
+
+	reg_map = dev->regs;
+
+	ad7124_8pmdz_fclk_logic_get(dev, &f_clk);
```
Comment (cristiantpop) [inline iio/iio_ad7124/iio_ad7124.c:639]: Check return

**Comment by adrimbarean** [comment]
V1: 
 - check return values of functions;
 - remove extra lines;
 - set default scan type for channels;
 - avoid init function by making a structure instance directly;
 - use iio_app following the iio_demo model;
 - update driver according to new library.

**Review by dbogdan — APPROVED** [review]

File: iio/iio_ad7124/iio_ad7124.c [hunk 8]
```diff
@@ -0,0 +1,849 @@
...
+
+/**
+ * @brief Read and display channel offset.
+ * @param device - Device driver descriptor.
+ * @param buf - Output buffer.
+ * @param len - Length of the input buffer (not used in this case).
+ * @param channel - IIO channel information.
+ * @return Number of bytes printed in the output buffer, or negative error code.
+ */
+static ssize_t ad1724_iio_read_offset_chan(void *device, char *buf, size_t len,
```
Comment (mchindri) [inline iio/iio_ad7124/iio_ad7124.c:151]: typo in name: ad7124

File: iio/iio_ad7124/iio_ad7124.c [hunk 9]
```diff
@@ -0,0 +1,849 @@
...
+{
+	struct ad7124_dev	*desc = (struct ad7124_dev *)device;
+	uint32_t		value;
+	int32_t ret;
+
+	ret = ad7124_read_register(desc,
+				   &desc->regs[(AD7124_OFFS0_REG + channel->ch_num)]);
+	if (ret != SUCCESS)
+		return ret;
+	value = desc->regs[(AD7124_OFFS0_REG + channel->ch_num)].value;
```
Comment (mchindri) [inline iio/iio_ad7124/iio_ad7124.c:162]: I think fields from inside desc should pe only accesed inside the drivere that created that structure. So my approach will be to create in the driver a function ad7124_read_offset_chan . 
Same for other the remaining functions

Comment (mchindri) [reply iio/iio_ad7124/iio_ad7124.c:162]: And then you can do a switch calling that function like this: https://github.com/analogdevicesinc/no-OS/blob/05cd411971cb9368179cdadbf695d6a95f404ba3/drivers/platform/aducm3029/iio_aducm3029.c#L222

**Comment by mchindri** [comment]
Move iio files in drivers/adc/ad7124
https://github.com/analogdevicesinc/no-OS/pull/889#discussion_r552625001

File: iio/iio_ad7124/iio_ad7124.c (hunk 8 above)
Comment (adrimbarean) [reply iio/iio_ad7124/iio_ad7124.c:151]: Will do.

File: iio/iio_ad7124/iio_ad7124.c [hunk 10]
```diff
@@ -0,0 +1,849 @@
...
+		.name = "scale",
+		.priv = 0,
+		.show = ad1724_iio_read_scale_chan,
+		.store = ad1724_iio_change_scale_chan
+	},
+	END_ATTRIBUTES_ARRAY
+};
+
+struct scan_type ad7124_iio_scan_type = {
+	.sign = 'i',
```
[the remark this answers is not in the document]
Comment (adrimbarean) [reply iio/iio_ad7124/iio_ad7124.c:112]: Will fix.

File: iio/iio_ad7124/iio_ad7124.c [hunk 11]
```diff
@@ -0,0 +1,849 @@
...
+	AD7124_IIO_CHANN_DEF("ch0", 0, 1),
+	AD7124_IIO_CHANN_DEF("ch1", 2, 3),
+	AD7124_IIO_CHANN_DEF("ch2", 4, 5),
+	AD7124_IIO_CHANN_DEF("ch3", 6, 7),
+	AD7124_IIO_CHANN_DEF("ch4", 8, 9),
+	AD7124_IIO_CHANN_DEF("ch5", 10, 11),
+	AD7124_IIO_CHANN_DEF("ch6", 12, 13),
+	AD7124_IIO_CHANN_DEF("ch7", 14, 15)
+};
+
```
[the remark this answers is not in the document]
Comment (adrimbarean) [reply iio/iio_ad7124/iio_ad7124.c:142]: I can do this, but it seems that the general rule is to define variables in a source file and share it with extern in header. Do we want to make our own rule?

**Comment by adrimbarean** [comment]
V2:
 - correct scan_type sign value;
 - correct ad1724 -> ad7124 typo;
 - implement requested changes in the device driver and act accordingly: https://github.com/analogdevicesinc/no-OS/pull/904

**Review by mchindri — APPROVED** [review]
Beside the iio definitions are in the .c file, LGTM

**Comment by dbogdan** [comment]
Where was ad7124_read_register_wrap() implemented? I think Pipelines' complaints are valid.

**Comment by mchindri** [comment]
> Where was ad7124_read/write_register_wrap() implemented? I think Pipelines' complaints are valid.

In #904. Maybe that Pr should be merged first or add it as a commit here

**Comment by adrimbarean** [comment]
V3:
 - update the code to work with new ad7124 driver https://github.com/analogdevicesinc/no-OS/pull/904;

**Comment by adrimbarean** [comment]
V4:
 - fix astyle.

**Pull request merged**